### ISSUE TITLE
INSERT/DELETE WITH semantics vs FROM semantics

### ISSUE THREAD

**Issue by rubensworks** [joachimvh/SPARQLAlgebra.js#78]
_This issue is similar to https://github.com/joachimvh/SPARQLAlgebra.js/issues/77_

After fixing Comunica's implementation of `FROM` based on the conclusion in https://github.com/w3c/rdf-tests/issues/72#issuecomment-800217001, the two respective tests now pass.
However, two other (very similar) tests started failing:

* http://www.w3.org/2009/sparql/docs/tests/data-sparql11/delete/manifest#dawg-delete-with-02
* http://www.w3.org/2009/sparql/docs/tests/data-sparql11/delete/manifest#dawg-delete-with-06

These tests are similar to the ones before, but they use `WITH` instead of `USING`.
There seems to be a small difference in semantics between `USING` and `WITH`.
From the spec:

> Note that explicit GRAPH clauses override a WITH clause. WITH provides a fallback to specify a graph (different from the default graph) to use when one is not explicitly stipulated via GRAPH.

As shown before, explicit `GRAPH` clauses _do not_ override `USING` clauses.

However, this lib also makes use of the `FROM` clause to wrap the `WHERE` clause, which makes a query such as the following not function as expected: (the transformation of `DELETE`/`INSERT` seems correct though)

```sparql
PREFIX     : <http://example.org/>
PREFIX foaf: <http://xmlns.com/foaf/0.1/>

WITH <http://example.org/g2>
DELETE
{
  GRAPH <http://example.org/g1> { ?s ?p ?o }
}
WHERE
{
  GRAPH <http://example.org/g1> { :a foaf:knows ?s .
                                  ?s ?p ?o }
}
```

Please let me know if you think I'm missing something again :-)
But if I'm correct, then we may consider exposing `WITH` explicitly on algebra level, but there may be other less-invasive solutions.

**Comment by joachimvh** [comment]
Well I guess this is a lie 😉 https://github.com/joachimvh/SPARQLAlgebra.js/blob/f62a8b5efc948df66898b3a369ec35f059ab7a70/lib/sparqlAlgebra.ts#L688-L689

Not sure why I did it like this. Am I wrong in saying that we could just wrap the WHERE block in a GRAPH statement since that is what the spec seems to say?

> That is, a WITH clause may be viewed as syntactic sugar for wrapping both the QuadPatterns in subsequent DELETE and INSERT clauses, and likewise the GroupGraphPattern in the subsequent WHERE clause into GRAPH patterns. This can be used to avoid refering to the same graph multiple times in a single operation.

**Comment by rubensworks** [comment]
Yeah, the spec seems to be a bit confusing in that part.

When they talk about that equivalence, they also mention this exception right afterwards:

> To illustrate the use of the WITH clause, an operation of the general form:

> WITH <g1> DELETE { a b c } INSERT { x y z } WHERE { ... }

> is considered equivalent to:

> DELETE { GRAPH <g1> { a b c } } INSERT { GRAPH <g1> { x y z } } USING <g1> WHERE { ... }

> Note that explicit GRAPH clauses override a WITH clause. WITH provides a fallback to specify a graph (different from the default graph) to use when one is not explicitly stipulated via GRAPH.

So they seem to suggest that WITH can just be considered syntactical sugar, but not really though :P

**Comment by joachimvh** [comment]
But wait, that example is exactly what SparqlAlgebra does right now, so that explains why it's implemented as it is since USING and FROM are equivalent. So either that example is wrong, or the test results are wrong?

**Comment by rubensworks** [comment]
Well, they do mention one exception for this example:

> Note that explicit GRAPH clauses override a WITH clause. WITH provides a fallback to specify a graph (different from the default graph) to use when one is not explicitly stipulated via GRAPH.

So the example+exception seems to match what is expected in the tests.

I guess the example is mainly useful for people writing SPARQL queries, but not for implementers, as queries can not really be rewritten in that way (due to the exception).

**Comment by joachimvh** [comment]
Should be fixed by d0d0657ccda789d48dc7ab870e3432d7acf19882, feel free te re-open if I'm wrong.

**Comment by rubensworks** [comment]
Thanks, will check on my end soon.
